android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java

```diff
@@ -0,0 +1,123 @@
+package com.tavultesoft.kmea.util;
+
+import java.util.Map;
+import java.util.Scanner;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * A custom, lightweight INI parser designed initially for use with Keyman's .kmp and .inf files.
+ * @author jahorton
+ *
+ */
+public class INIParser {
+	public static Map<String, Map<String, List<String>>> parse(String _string) throws INIParseException {
+		Map<String, Map<String, List<String>>> sections = new HashMap<String, Map<String, List<String>>>();
+		
+		String[] linesArray = _string.split("\n");
+		ArrayList<String> lines = new ArrayList<String>(linesArray.length);
+		
+		// Perform comment & whitespace filtering.
+		for(int i=0; i < linesArray.length; i++) {
+			String line = linesArray[i].trim();
+			if(!(line.equals("") || line.charAt(0) == ';')) {
+				lines.add(line);
+			}
+		}
+		
+		// Allow a plain default section - if it goes unutilized, we can remove it later.
+		String currentSectionName = "";
+		Map<String, List<String>> currentSection = new HashMap<String, List<String>>();
+		sections.put(currentSectionName, currentSection);
+		
+		// Main parsing loop.
+		for(String line:lines) {
+			if(line.charAt(0) == '[') {
+				currentSectionName = unquote_helper(line, '[', ']', 
+						"A section definition is missing its closing bracket",
+						"Improper section header definition - extra characters detected");
+						
+				
+				if(sections.containsKey(currentSectionName)) {
+					throw new INIParseException("Section \"" + currentSectionName + "\" has a duplicate entry!");
+				}
+				
+				currentSection = new HashMap<String, List<String>>();
+				sections.put(currentSectionName, currentSection);
+			} else if(line.indexOf('=') != -1) {
+				int eq = line.indexOf('=');
+				
+				String key = line.substring(0, eq).trim();
+				String values = line.substring(eq + 1).trim();
+				
+				if(currentSection.containsKey(key)) {
+					throw new INIParseException("Section property duplicated for key \"" + key + "\" in section \"" + currentSectionName + "\".");
+				}
+				
+				currentSection.put(key, components(values));
+			}
+		}
+		
+		// If the default section is empty, remove it.
+		if(sections.get("").isEmpty()) {
+			sections.remove("");
+		}
+		
+		return sections;
+	}
+	
+	public static List<String> components (String text) throws INIParseException {
```

Comment:
I think we should avoid doing this in the `INIParser` class: technically the INI format is just name=value pairs. This way we can use the class on arbitrary INI files in the future if we had reason to.

Also some values in the .inf format may have "," which are part of their value, and not separating components, for example `[Keyboard0].Name` may have a value with a comma in it. Thus, this componentization should probably only be done for the entries that specifically require it, according to the kmp.inf spec

Reply:
> Thus, this componentization should probably only be done for the entries that specifically require it, according to the kmp.inf spec.

Got it, that'll be a part two thing.  Though... that leaves me wondering about this situation.

```
value="foo","bar"
value2="foobar"
```

How do we ensure we don't, well, FUBAR either value there?